### PR TITLE
Initial Swagger client generator

### PR DESCRIPTION
**Pull request by manik1337**
- [x] Base generation
- [x] Handle parameters
- [x] Validation

### CODE REVIEW

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/ErlangClientCodegen.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    @Override
+    public String toModelName(String name) {
+        return camelize(toModelFilename(name));
+    }
+
+    @Override
+    public String toOperationId(String operationId) {
+        // method name cannot use reserved keyword, e.g. return
+        if (isReservedWord(operationId)) {
+            LOGGER.warn(operationId + " (reserved word) cannot be used as method name. Renamed to " + camelize(sanitizeName("call_" + operationId)));
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/ErlangClientCodegen.java]: It looks like this log message is trying to lie to me. OperationId is going to be `underscore` after all.

File: modules/swagger-codegen/src/main/resources/erlang-client/api.mustache [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+{{#operations}}{{#operation}}
+-spec {{operationId}}(Headers::list(), Body::term()) ->
+  {ok, integer(), list(), client_ref()} | {ok, integer(), list()} | {error, term()}.
+
+{{operationId}}(Body, Headers) ->
+  Url = get_url({{basePathWithoutHost}}{{path}}),
+  PreparedBody = jsx:encode(Body),
+
+  {ok, Code, RespHeaders, ClientRef} = hackney:request({{httpMethod}}, Url, Headers, PreparedBody),
+  {ok, Code, RespHeaders, ClientRef}.
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/api.mustache]: It looks like this line can be deleted

File: modules/swagger-codegen/src/main/resources/erlang-client/app.src.mustache [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{application, {{packageName}}, [
+    {description, {{#appDescription}}"{{appDescription}}"{{/appDescription}}{{^appDescription}}"Swagger rest server library"{{/appDescription}}},
+    {vsn, "{{apiVersion}}"},
+    {registered, []},
+    {applications, [
+        kernel,
+        stdlib,
+        jsx,
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/app.src.mustache]: Wrong comma

File: modules/swagger-codegen/src/main/resources/erlang-client/utils.mustache [hunk 4]
```diff
@@ -0,0 +1,12 @@
+-module(utils).
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/utils.mustache]: We still need a prefix here

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api.mustache [hunk 5]
```diff
@@ -0,0 +1,16 @@
+-module({{classname}}).
+
+%% generated methods
+{{#operations}}{{#operation}}-export([{{operationId}}/3]).
+{{/operation}}{{/operations}}
+{{#operations}}{{#operation}}
+-spec {{operationId}}(Host::string(), Port::integer(), Params::map()) ->
+    {ok, integer(), list(), {ok, binary()}}.
+{{operationId}}(Host, Port, Params) ->
+    swagger_client_api_procession:process_request(
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api.mustache]: I don't think that hardcoding prefix in a template is a good idea

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 6]
```diff
@@ -0,0 +1,136 @@
+-module({{packageName}}_client_api_procession).
+
+-type operation_id() :: atom().
+
+-record(state, {
+    operation_id :: operation_id(),
+    validator_state :: jesse_state:state()
+}).
+
+-type state() :: state().
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: Will it compile?

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+-export([process_request/4]).
+
+-spec process_request(Method::atom(), OperationID::atom(), Url::string(), Params::map()) ->
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: I would suggest to export `operation_id` type and use always use it

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    {ok, integer(), list(), {ok, binary()}}.
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: Please consider adding some explanations for this values.

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+process_request(Method, OperationID, Url, Params) ->
+    State = #state{
+        operation_id = OperationID,
+        validator_state = prepare_validator()
+    },
+
+    {PreparedUrlParams, PreparedHeaderParams, PreparedBodyParams, PreparedQueryParams} =
+        handle_request_json(Params, State),
+
+    #{} = PreparedQueryParams,
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: Why do we have to check it here?

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    PreparedUrl = prepare_url(Url, PreparedUrlParams, PreparedQueryParams),
+    PreparedHeaders = maps:to_list(PreparedHeaderParams),
+    PreparedBody = jsx:encode(PreparedBodyParams),
+
+    hackney:request(Method, PreparedUrl, PreparedHeaders, PreparedBody, [with_body]).
+
+-spec handle_request_json(term(), state()) -> {ok, map(), map(), map(), map()} | {error, term()}.
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: This function actually never returns `{ok, ...}`

Comment (galaxie) [reply modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: I would add some names to these `term()`, `map()`

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        {error, Reason, _} ->
+            {error, Reason}
+    end.
+
+-spec populate_request(
+    OperationID :: term(),
+    RequestParams :: term(),
+    ValidatorState :: jesse_state:state()
+) ->
+    {ok, Params :: map()} |
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache:53]: It looks like this `Params` is more then just a map, but a map with a specific set of keys (e.g. `qs_val`). How about describing it somewhere?

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+
+get_swagger_path() ->
+    {ok, AppName} = application:get_application(?MODULE),
+    filename:join(swagger_client_api_utils:priv_dir(AppName), "swagger.json").
+
+-spec get_value(atom(), term(), term()) -> term().
+get_value(body, _Name, RequestParams) ->
+    Body = maps:get(body, RequestParams),
+    Value = Body,
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: Looks a bit useless, but who am I to judge

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+-export([request_param_info/2]).
+-export([validate/4]).
+-export([prepare_body/1]).
+
+-type operation_id() :: atom().
+-type request_param() :: atom().
+
+
+-type request_context() :: #{
+    auth_context => AuthContext :: swagger_auth:context(),
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache]: Where did `swagger_auth` come from?

**Review by galaxie — CHANGES_REQUESTED** [review]

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 14]
```diff
@@ -45,42 +45,42 @@ handle_request_json(
...
+    Params = {{packageName}}_client_api_validation:request_params(OperationID),
     populate_request_params(OperationID, Params, RequestParams, ValidatorState).
 
 populate_request_params(_, [], RequestParams, _) ->
     {ok, RequestParams};
 populate_request_params(OperationID, [FieldParams | T], RequestParams, ValidatorState) ->
     case populate_request_param(OperationID, FieldParams, RequestParams, ValidatorState) of
         {ok, _K, _V, RequestParams} ->
-            % populate_request_params(OperationID, T, genlib_map:deepput(swagger_client_api_utils:to_binary(K), V, RequestParams), ValidatorState);
+            % populate_request_params(OperationID, T, genlib_map:deepput({{packageName}}_client_api_utils:to_binary(K), V, RequestParams), ValidatorState);
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: I would do something with this commented string (like remove?)

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+        handle_request_json(Params, State),
+
+    PreparedUrl = prepare_url(Url, PreparedUrlParams, PreparedQueryParams),
+    PreparedHeaders = maps:to_list(PreparedHeaderParams),
+    PreparedBody = jsx:encode(PreparedBodyParams),
+
+    hackney:request(Method, PreparedUrl, PreparedHeaders, PreparedBody, [with_body]).
+
+-spec handle_request_json(RequestParams::params(), State::state()) ->
+    {UrlParams::map(), HeaderParams::map(), BodyParams::map(), QueryParams::map()} | {error, term()}.
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: I would recommend to use `{ok, ....} | {error, ...}` response or `term() | no_return()`

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api.mustache [hunk 16]
```diff
@@ -0,0 +1,16 @@
+-module({{classname}}).
+
+%% generated methods
+{{#operations}}{{#operation}}-export([{{operationId}}/3]).
+{{/operation}}{{/operations}}
+{{#operations}}{{#operation}}
+-spec {{operationId}}(Host::string(), Port::integer(), Params::map()) ->
+    {ok, integer(), list(), {ok, binary()}}.
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api.mustache]: Any comments about these `integer()`, `list()` etc?

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+populate_request(OperationID, RequestParams, ValidatorState) ->
+    Params = {{packageName}}_client_api_validation:request_params(OperationID),
+    populate_request_params(OperationID, Params, RequestParams, ValidatorState).
+
+populate_request_params(_, [], RequestParams, _) ->
+    {ok, RequestParams};
+populate_request_params(OperationID, [FieldParams | T], RequestParams, ValidatorState) ->
+    case populate_request_param(OperationID, FieldParams, RequestParams, ValidatorState) of
+        {ok, _K, _V, RequestParams} ->
+            % populate_request_params(OperationID, T, genlib_map:deepput({{packageName}}_client_api_utils:to_binary(K), V, RequestParams), ValidatorState);
```
Comment (galaxie) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_procession.mustache]: How about removing it?

**Review by galaxie — CHANGES_REQUESTED** [review]

**Review by galaxie — APPROVED** [review]
:shipit:

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+    {enum, [atom()]} |
+    {max, Max :: number()} |
+    {exclusive_max, Max :: number()} |
+    {min, Min :: number()} |
+    {exclusive_min, Min :: number()} |
+    {max_length, MaxLength :: integer()} |
+    {min_length, MaxLength :: integer()} |
+    {pattern, Pattern :: string()} |
+    schema |
+    required |
```
Comment (keynslug) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache:51]: Тут спека тогда кривая, но это и в `erlang-server` так, возможно стоит и там поправить в рамках мини-pr.

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache [hunk 19]
```diff
@@ -0,0 +1,274 @@
+-module({{packageName}}_client_api_validation).
```
Comment (keynslug) [inline modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache:1]: А можно ли переиспользовать шаблоны от других таргетов, чтобы не копипастить?

File: modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache (hunk 18 above)
Comment (manik1337) [reply modules/swagger-codegen/src/main/resources/erlang-client/client_api_validation.mustache:51]: 👍 Сделаю в виде отдельного pr

**Pull request merged**